File: 0112/bfs.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if root is None:
            return False
        q = deque()
        q.append((root, 0))
        while len(q) > 0:
            node, prev_sum = q.popleft()
            curr_sum = prev_sum + node.val
            if not node.left and not node.right and curr_sum == sum:
                return True
            if node.left:
                q.append((node.left, curr_sum))
            if node.right:
                q.append((node.right, curr_sum))
        return False


if __name__ == '__main__':
    sol = Solution()

    root = TreeNode(5)
    root.left = TreeNode(4)
    root.left.left = TreeNode(11)
    root.left.left.left = TreeNode(7)
    root.left.left.right = TreeNode(2)
    root.right = TreeNode(8)
    root.right.left = TreeNode(13)
    root.right.right = TreeNode(4)
    root.right.right.right = TreeNode(1)
    assert sol.hasPathSum(root, 22)

File: 0637/bfs.cpp
#include <gtest/gtest.h>

#include <cstdlib>
#include <queue>
#include <string>
#include <vector>

#include "leetcode/tree.hpp"

using namespace std;
using namespace LeetCode;

class Solution {
 public:
  vector<double> averageOfLevels(TreeNode* root) {
    vector<double> ret;
    if (!root) {
      return ret;
    }
    queue<pair<TreeNode*, int>> q;
    q.push({root, 1});
    long sum = 0;
    int count = 0;
    int curr_level = 1;
    while (!q.empty()) {
      TreeNode* node = q.front().first;
      int level = q.front().second;
      q.pop();
      if (level != curr_level) {
        ret.push_back(static_cast<double>(sum) / count);
        curr_level = level;
        sum = node->val;
        count = 1;
      } else {
        sum += node->val;
        count++;
      }
      if (node->left) {
        q.push({node->left, level + 1});
      }
      if (node->right) {
        q.push({node->right, level + 1});
      }
    }
    ret.push_back(static_cast<double>(sum) / count);
    return ret;
  }
};

int main() {
  Solution sol;
  TreeNode* root;
  vector<double> expected;

  root = CreateTreeNode("3 9 20 # # 15 7"), expected = {3.0, 14.5, 11.0};
  EXPECT_EQ(expected, sol.averageOfLevels(root));
  DeleteTreeNode(root);

  root = nullptr, expected = {};
  EXPECT_EQ(expected, sol.averageOfLevels(root));

  exit(EXIT_SUCCESS);
}
